hr_timesheet_sheet/models/hr_timesheet_sheet.py

```diff
@@ -559,8 +641,46 @@ def onchange_unit_amount(self):
         if not sheet:
             return {'warning': {
                 'title': _("Warning"),
-                'message': _("Save the Timesheet Sheet first.")
+                'message': _("Save the Timesheet Sheet first."),
             }}
+        return sheet
+
+
+class SheetNewAnalyticLine(models.TransientModel):
+    _name = 'hr_timesheet.sheet.new.analytic.line'
+    _description = 'Timesheet Sheet New Analytic Line'
+
+    sheet_id = fields.Many2one(
```

Comment:
The two transient models have the most of the fields in common. You can move them in an abstract model. This will reduce the lines of code.

Create a new abstract model:

```python
class AbstractSheetLine(models.AbstractModel):
    _name = 'hr_timesheet.sheet.line.abstract'
    _description = 'Abstract Timesheet Sheet Line'

    sheet_id = fields.Many2one('hr_timesheet.sheet', ondelete='cascade')
    date = fields.Date()
    project_id = fields.Many2one('project.project', string='Project')
    task_id = fields.Many2one('project.task', string='Task')
    unit_amount = fields.Float(string="Quantity", default=0.0)
    company_id = fields.Many2one('res.company', string='Company')
    employee_id = fields.Many2one('hr.employee', string='Employee')
```

And then: 

```python
class SheetLine(models.TransientModel):
    _name = 'hr_timesheet.sheet.line'
    _inherit = 'hr_timesheet.sheet.line.abstract'
    _description = 'Timesheet Sheet Line'

    value_x = fields.Char('Date Name')
    value_y = fields.Char('Project Name')
```

and:

```python
class SheetNewAnalyticLine(models.TransientModel):
    _name = 'hr_timesheet.sheet.new.analytic.line'
    _inherit = 'hr_timesheet.sheet.line.abstract'
    _description = 'Timesheet Sheet New Analytic Line'
```

Reply:
thanks, all comments attended :+1: